src/coprocessor/coprocessor.rs

```diff
@@ -0,0 +1,1142 @@
+                future::result(handler_builder.call_box((snapshot, &tracker.req_ctx)))
+                    .map(|handler| (tracker, handler))
+            });
+
+        stream::futures_ordered(vec![tracker_and_handler_future])
+            .map(|(mut tracker, handler)| {
+                tracker.on_begin_all_items();
+
+                // The state is `Option<(tracker, handler, finished)>`, `None` indicates finished.
+                // For every stream item except the last one, the type is `Either::Left(Response)`.
+                // For last stream item, the type is `Either::Right(Tracker)` so that we can do
+                // more things for tracker later.
+                let initial_state = Some((tracker, handler, false));
+                stream::unfold(initial_state, |state| {
+                    match state {
+                        Some((mut tracker, mut handler, finished)) => {
+                            if finished {
+                                // Emit tracker as the last item.
+                                let yielded = Either::Right(tracker);
+                                let next_state = None;
+                                return Some(Ok((yielded, next_state)));
+                            }
+
+                            // There are future items
+                            tracker.on_begin_item();
+
+                            let result = handler.handle_streaming_request();
+                            let exec_metrics = {
+                                let mut metrics = ExecutorMetrics::default();
+                                handler.collect_metrics_into(&mut metrics);
+                                metrics
+                            };
+
+                            tracker.on_finish_item(Some(exec_metrics));
+                            let exec_details = tracker.get_item_exec_details();
+
+                            let (mut resp, finished) = match result {
+                                Err(e) => (make_error_response(e), true),
+                                Ok((None, _)) => {
+                                    let yielded = Either::Right(tracker);
+                                    let next_state = None;
+                                    return Some(Ok((yielded, next_state)));
+                                }
+                                Ok((Some(resp), finished)) => (resp, finished),
+                            };
+                            resp.set_exec_details(exec_details);
+
+                            let yielded = Either::Left(resp);
+                            let next_state = Some((tracker, handler, finished));
+                            Some(Ok((yielded, next_state)))
+                        }
+                        None => {
+                            // Finished
+                            None
+                        }
+                    }
+                }).filter_map(|resp_or_tracker| match resp_or_tracker {
```

Comment:
Is `Either` necessary here？Why not just handle the `tracker` directly in `stream::unfold`.

Reply:
It is not necessary, only to be more readable.